src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelProxyDiscoveryManager.cs

```diff
@@ -145,27 +144,27 @@ public bool HandlePartialDiscoveryComplete(IProxyDiscoveryManager proxyDiscovery
                 return true;
             }
 
-            // Discovery is not complete.
-            // First, clean up the used proxy discovery manager if the last run was aborted
-            // or this run doesn't support shared hosts (netcore tests)
-            if (!this.SharedHosts || isAborted)
+            /*  Discovery is not complete.
+                Now when both.net framework and.net core projects can run in parallel
+                we should clear manager and create new one for both cases.
+                Otherwise `proxyDiscoveryManager` instance is alredy closed by now and it will give exception
+                when trying to do some operation on it.
+            */
+            if (EqtTrace.IsVerboseEnabled)
             {
-                if (EqtTrace.IsVerboseEnabled)
-                {
-                    EqtTrace.Verbose("ParallelProxyDiscoveryManager: HandlePartialDiscoveryComplete: Replace discovery manager. Shared: {0}, Aborted: {1}.", this.SharedHosts, isAborted);
-                }
+                EqtTrace.Verbose("ParallelProxyDiscoveryManager: HandlePartialDiscoveryComplete: Replace discovery manager. Shared: {0}, Aborted: {1}.", this.SharedHosts, isAborted);
+            }
 
-                this.RemoveManager(proxyDiscoveryManager);
+            this.RemoveManager(proxyDiscoveryManager);
 
-                proxyDiscoveryManager = this.CreateNewConcurrentManager();
```

Comment:
We should probably take advantage of the fact that .net fw testhost is shared and re-use it instead of killing it in the parallel discovery/execution case.

E.g.: `parallelLevel = 4`, .net fw test assmeblies = 8.

What happens now is that we spawn 4 testhosts, discover 4 .net fw test assmeblies, kill the 4 testhosts and then spawn another 4 testhosts to do the same for the remaining 4 test assemblies.

Reply:
Actually, after discovery 4 test assemblies, it will spawn next 4 testhosts one by one, not simultaniously as the first time.
So it will spawn 4 testhosts and when some testhost finishes discovering it will start new one for the next source.

As we discussed, I just created new issue for that idea : #3193.
Feel free to add or edit description for the issue.